Implement conditional stripping of type descriptors, protocols and protocol conformances via !llvm.used.conditional

kubamracek opened the pull request:
A new LLVM IR affordance that allows expressing conditions under which globals
can be removed/dropped (even when marked with @llvm.used) is being discussed at:

- <https://reviews.llvm.org/D104496>
- <https://lists.llvm.org/pipermail/llvm-dev/2021-September/152656.html>

This is a preliminary implementation that marks runtime lookup records (namely
protocol records, type descriptors records and protocol conformance records)
with the !llvm.used.conditional descriptors. That allows link-time / LTO-time
removal of these records (by GlobalDCE) based on whether they're actually used
within the linkage unit. Effectively, this allows libraries that have a limited
and known set of clients, to be optimized against the client at LTO time, and
significantly reduce the code size of that library.

Parts of the implementation:
- New -conditional-runtime-records frontend flag to enable using !llvm.used.conditional
- IRGen code that emits these records can now emit these either as a single contiguous
  array (asContiguousArray = true, the old way), which is used for JIT mode, or
  as indivial globals (asContiguousArray = false), which is necessary for the
  !llvm.used.conditional stripping to work.
- When records are emitted as individual globals, they have new names of
  "\01l_protocol_" + mangled name of the protocol descriptor, and similarly for
  other records.
- Fixed existing tests to account for individual records instead of a single array
- Added an IR level test, and an end-to-end execution test to demonstrate that
  the !llvm.used.conditional-based stripping actually works.

kubamracek commented:
@swift-ci please test

eeckstein commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -3649,34 +3651,62 @@ IRGenModule::emitDirectRelativeReference(llvm::Constant *target,
-  }
+static void
+appendLLVMUsedConditionalEntry(IRGenModule &IGM, llvm::GlobalVariable *var,
+                               const ConformanceDescription &record) {
```
Can you add a comment saying that `var` depends on record.protocol and record.type

eeckstein commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -3697,12 +3727,60 @@ llvm::Constant *IRGenModule::emitSwiftProtocols() {
+    auto var = new llvm::GlobalVariable(
+        Module, ProtocolRecordTy, /*isConstant*/ true,
+        llvm::GlobalValue::PrivateLinkage, /*initializer*/ nullptr,
+        "\x01l_protocol_" + link.getName());
```
That naming convention breaks the ability to demangle the symbol. Also, it adds a lot of symbol name string data.
I'd prefer to add a mangling scheme for these variables.

eeckstein commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -3697,12 +3727,60 @@ llvm::Constant *IRGenModule::emitSwiftProtocols() {
+    disableAddressSanitizer(*this, var);
+    addUsedGlobal(var);
+
+    if (IRGen.Opts.ConditionalRuntimeRecords) {
```
Can you add a comment describing what depends on what (it's not obvious to me).

eeckstein commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -3758,17 +3817,70 @@ llvm::Constant *IRGenModule::emitProtocolConformances() {
+    auto var = new llvm::GlobalVariable(
+        Module, RelativeAddressTy, /*isConstant*/ true,
+        llvm::GlobalValue::PrivateLinkage, /*initializer*/ nullptr,
+        "\x01l_protocol_conformance_" + link.getName());
```
Same here: it's better to add a mangled name

eeckstein commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -3758,17 +3817,70 @@ llvm::Constant *IRGenModule::emitProtocolConformances() {
+    disableAddressSanitizer(*this, var);
+    addUsedGlobal(var);
+
+    if (IRGen.Opts.ConditionalRuntimeRecords) {
```
Again: please add a comment describing what depends on what

eeckstein commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -3788,54 +3900,77 @@ llvm::Constant *IRGenModule::emitTypeMetadataRecords() {
+    auto var = new llvm::GlobalVariable(
+        Module, TypeMetadataRecordTy, /*isConstant*/ true,
+        llvm::GlobalValue::PrivateLinkage, /*initializer*/ nullptr,
+        "\x01l_type_metadata_" + ref.getValue()->getName());
```
again: mangling

eeckstein commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -3788,54 +3900,77 @@ llvm::Constant *IRGenModule::emitTypeMetadataRecords() {
+    addUsedGlobal(var);
+
+    if (IRGen.Opts.ConditionalRuntimeRecords) {
+      appendLLVMUsedConditionalEntry(*this, var, ref.getValue());
```
again: comment

eeckstein reviewed (commented):
basically LGTM.
See my comments

kubamracek commented:
@swift-ci please test

kubamracek commented:
@swift-ci please test

kubamracek committed:
For individual runtime records for types/protocols/conformances, add and use new mangling suffixes

kubamracek committed:
Implement conditional stripping of type descriptors, protocols and protocol conformances via !llvm.used.conditional

A new LLVM IR affordance that allows expressing conditions under which globals
can be removed/dropped (even when marked with @llvm.used) is being discussed at:

- <https://reviews.llvm.org/D104496>
- <https://lists.llvm.org/pipermail/llvm-dev/2021-September/152656.html>

This is a preliminary implementation that marks runtime lookup records (namely
protocol records, type descriptors records and protocol conformance records)
with the !llvm.used.conditional descriptors. That allows link-time / LTO-time
removal of these records (by GlobalDCE) based on whether they're actually used
within the linkage unit. Effectively, this allows libraries that have a limited
and known set of clients, to be optimized against the client at LTO time, and
significantly reduce the code size of that library.

Parts of the implementation:
- New -conditional-runtime-records frontend flag to enable using !llvm.used.conditional
- IRGen code that emits these records can now emit these either as a single contiguous
  array (asContiguousArray = true, the old way), which is used for JIT mode, or
  as indivial globals (asContiguousArray = false), which is necessary for the
  !llvm.used.conditional stripping to work.
- When records are emitted as individual globals, they have new names of
  "\01l_protocol_" + mangled name of the protocol descriptor, and similarly for
  other records.
- Fixed existing tests to account for individual records instead of a single array
- Added an IR level test, and an end-to-end execution test to demonstrate that
  the !llvm.used.conditional-based stripping actually works.

kubamracek commented:
@swift-ci please test

kubamracek commented:
@swift-ci please test Windows platform

eeckstein reviewed (approved):
LGTM!

@aschwaighofer can to please take a look, too?

aschwaighofer reviewed (approved):
lgtm

kubamracek commented:
Thanks for the review!
